### PR TITLE
Device Code Flow into Vapor OAuth

### PR DESCRIPTION
**Pull request by vamsii777**
This pull request introduces the integration of the Device Code Flow into the Vapor OAuth, enhancing the existing OAuth2 Provider capabilities.

### Changes:
- Added Device Code Flow support following RFC 8628.
- Implemented `DeviceCodeTokenHandler` to manage device code token requests.
- Updated test cases to ensure the correct handling of device code expiration.

### Impact:
This integration allows applications to utilize the Device Code Flow for authorization, providing a more secure and user-friendly way to authenticate devices without a browser or other user agent.

### Testing:
All test cases have been updated and passed, ensuring that the new feature is working as expected and does not break existing functionality.

Please review the changes and provide feedback. Thank you!

### CODE REVIEW

**Comment by mynona** [comment]
@0xTim  @marius-se  Will this request be merged? Would also be great :-)

**Comment by Aquaa7** [comment]
Would be great if it’s merged! But is this package abandoned?

**Comment by marius-se** [comment]
I'll look into this early next week! Sorry for the late reply, this project is **not** abandoned!

File: Sources/VaporOAuth/DefaultImplementations/EmptyCodeManager.swift [hunk 1]
```diff
@@ -5,14 +5,27 @@ public struct EmptyCodeManager: CodeManager {
         return nil
     }
 
+    // Updated to include PKCE parameters
```
Comment (marius-se) [inline Sources/VaporOAuth/DefaultImplementations/EmptyCodeManager.swift]: no need for this comment here

File: Sources/VaporOAuth/DefaultImplementations/StaticClientRetriever.swift [hunk 2]
```diff
@@ -1,12 +1,11 @@
-public struct StaticClientRetriever: ClientRetriever {
+import Vapor
```
Comment (marius-se) [inline Sources/VaporOAuth/DefaultImplementations/StaticClientRetriever.swift]: import not needed

File: Sources/VaporOAuth/DefaultImplementations/StaticClientRetriever.swift [hunk 3]
```diff
@@ -1,12 +1,11 @@
...
 
-    let clients: [String: OAuthClient]
+public actor StaticClientRetriever: ClientRetriever {
```
Comment (marius-se) [inline Sources/VaporOAuth/DefaultImplementations/StaticClientRetriever.swift]: is there a reason why you turned this into an actor? I don't think we need this, but I might be wrong :)

File: Sources/VaporOAuth/Helper/OAuthHelper+remote.swift [hunk 4]
```diff
@@ -1,62 +1,79 @@
 import Vapor
 
+
+actor RemoteTokenResponseActor {
```
Comment (marius-se) [inline Sources/VaporOAuth/Helper/OAuthHelper+remote.swift:4]: same here, why an actor? structs are way cheaper!

File: Sources/VaporOAuth/Helper/OAuthHelper+remote.swift [hunk 5]
```diff
@@ -1,62 +1,79 @@
...
+    func hasTokenResponse() async -> Bool {
+        return remoteTokenResponse != nil
+    }
+
+    func getRemoteTokenResponse() async throws -> RemoteTokenResponse {
+        guard let response = remoteTokenResponse else {
+            throw Abort(.internalServerError)
+        }
+        return response
+    }
```
Comment (marius-se) [inline Sources/VaporOAuth/Helper/OAuthHelper+remote.swift]: remove`async` from function signature

File: Sources/VaporOAuth/Helper/OAuthHelper+remote.swift [hunk 6]
```diff
@@ -1,62 +1,79 @@
...
+actor RemoteTokenResponseActor {
+    var remoteTokenResponse: RemoteTokenResponse?
+
+    func setRemoteTokenResponse(_ response: RemoteTokenResponse) {
+        self.remoteTokenResponse = response
+    }
+
+    func hasTokenResponse() async -> Bool {
+        return remoteTokenResponse != nil
+    }
```
Comment (marius-se) [inline Sources/VaporOAuth/Helper/OAuthHelper+remote.swift]: remove `async` from function signature

File: Sources/VaporOAuth/Models/OAuthDeviceCode.swift [hunk 7]
```diff
@@ -0,0 +1,39 @@
+//
+//  OAuthDeviceCode.swift
+//
+//
+//  Created by Vamsi Madduluri on 24/08/23.
+//
+
```
Comment (marius-se) [inline Sources/VaporOAuth/Models/OAuthDeviceCode.swift]: remove these please :)

File: Sources/VaporOAuth/Models/OAuthDeviceCode.swift [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+import Foundation
+
+public final class OAuthDeviceCode {
```
Comment (marius-se) [inline Sources/VaporOAuth/Models/OAuthDeviceCode.swift]: can we turn this into a struct?

File: Sources/VaporOAuth/Models/OAuthDiscoveryDocument.swift [hunk 9]
```diff
@@ -0,0 +1,82 @@
+import Vapor
+
+public final class OAuthDiscoveryDocument: Content {
```
Comment (marius-se) [inline Sources/VaporOAuth/Models/OAuthDiscoveryDocument.swift]: same here (`struct`?)

File: Sources/VaporOAuth/Protocols/TokenManager.swift [hunk 10]
```diff
@@ -1,21 +1,47 @@
 import Vapor
 
-public protocol TokenManager {
-    func generateAccessRefreshTokens(
+public protocol TokenManager: Sendable {
+    // Generates access, refresh, and ID tokens. Should be called after successful authentication.
```
Comment (marius-se) [inline Sources/VaporOAuth/Protocols/TokenManager.swift]: let's turn these into `///` documenation comments. They're helpful and we want people to see them!

**Review by marius-se — CHANGES_REQUESTED** [review]
I didn't manage to go through the entire PR yet, but it's looking good so far! Just a few nits here and there :)

File: Sources/VaporOAuth/Helper/OAuthHelper+remote.swift (hunk 4 above)
Comment (vamsii777) [reply Sources/VaporOAuth/Helper/OAuthHelper+remote.swift:4]: Hi @marius-se, an actor is the best choice for our scenario, as it ensures thread-safe, concurrent access to shared mutable state, aligning well with Swift's concurrency model. While a class offers shared state, it lacks inherent thread safety. A struct, being a value type, isn't suited for shared, mutable state in concurrent contexts.

File: Sources/VaporOAuth/DefaultImplementations/EmptyCodeManager.swift (hunk 1 above)
Comment (vamsii777) [reply Sources/VaporOAuth/DefaultImplementations/EmptyCodeManager.swift]: Done

File: Sources/VaporOAuth/DefaultImplementations/StaticClientRetriever.swift (hunk 2 above)
Comment (vamsii777) [reply Sources/VaporOAuth/DefaultImplementations/StaticClientRetriever.swift]: Done

File: Sources/VaporOAuth/Helper/OAuthHelper+remote.swift (hunk 6 above)
Comment (vamsii777) [reply Sources/VaporOAuth/Helper/OAuthHelper+remote.swift]: Done

File: Sources/VaporOAuth/Helper/OAuthHelper+remote.swift (hunk 5 above)
Comment (vamsii777) [reply Sources/VaporOAuth/Helper/OAuthHelper+remote.swift]: Done

File: Sources/VaporOAuth/Models/OAuthDeviceCode.swift (hunk 7 above)
Comment (vamsii777) [reply Sources/VaporOAuth/Models/OAuthDeviceCode.swift]: Done

File: Sources/VaporOAuth/Protocols/TokenManager.swift (hunk 10 above)
Comment (vamsii777) [reply Sources/VaporOAuth/Protocols/TokenManager.swift]: Done

File: Sources/VaporOAuth/Models/OAuthDiscoveryDocument.swift (hunk 9 above)
Comment (vamsii777) [reply Sources/VaporOAuth/Models/OAuthDiscoveryDocument.swift]: For all the models or `OAuthDeviceCode` and `OAuthDiscoveryDocument`?

File: Sources/VaporOAuth/Helper/OAuthHelper+remote.swift (hunk 4 above)
Comment (marius-se) [reply Sources/VaporOAuth/Helper/OAuthHelper+remote.swift:4]: Right, sorry my question wasn't very precise, let me rephrase it: Why do we need (shared) mutable state here?

File: Sources/VaporOAuth/Models/OAuthDiscoveryDocument.swift (hunk 9 above)
Comment (marius-se) [reply Sources/VaporOAuth/Models/OAuthDiscoveryDocument.swift]: For all models if possible. As long as these types don't have mutable state there is no reason for them to be `class`es.

File: Sources/VaporOAuth/DefaultImplementations/StaticClientRetriever.swift (hunk 3 above)
Comment (vamsii777) [reply Sources/VaporOAuth/DefaultImplementations/StaticClientRetriever.swift]: No, it's just hmm... went with the flow I guess my bad. I've modified it with the use of struct.

File: Sources/VaporOAuth/Models/OAuthDeviceCode.swift (hunk 8 above)
Comment (vamsii777) [reply Sources/VaporOAuth/Models/OAuthDeviceCode.swift]: Done

File: Sources/VaporOAuth/Models/OAuthDiscoveryDocument.swift (hunk 9 above)
Comment (vamsii777) [reply Sources/VaporOAuth/Models/OAuthDiscoveryDocument.swift]: Done

File: Sources/VaporOAuth/Helper/OAuthHelper+remote.swift (hunk 4 above)
Comment (vamsii777) [reply Sources/VaporOAuth/Helper/OAuthHelper+remote.swift:4]: Yes, so the shared mutable state is needed for managing OAuth token responses across concurrent operations, ensuring updates and synchronization of token data in a multi-threaded environment.

File: Sources/VaporOAuth/DefaultImplementations/StaticClientRetriever.swift (hunk 3 above)
Comment (marius-se) [reply Sources/VaporOAuth/DefaultImplementations/StaticClientRetriever.swift]: awesome, thanks!

**Comment by 0xTim** [comment]
@vamsii777 I haven't done a full review of this yet but I have some initial feedback/queries. First, this is a pretty huge PR and ideally would be broken up so we could discuss some of the changes in isolation. However, some of the immediate things that I've spotted:
* I'm not 100% comfortable with dropping support for 5.8 and 5.7 just yet unless we have a very good reason
* We're annotating lots of stuff with Sendable, which is good, but should turn on complete concurrency checking
* The main issue - there's lots of stuff in the PR that's related to OpenID. That's got nothing to do with RFC 8628 and is very likely out of scope for this library. OpenID is a can of worms I don't particularly want to tackle in this library and would ideally be done with a library built on top of this one. None of it's required for device code flow right?

**Comment by Aquaa7** [comment]
Hi @0xTim,

Thank you for your initial review and feedback. I agree with your observations and would like to address the points you’ve raised:

1. Support for Swift 5.7 and 5.8: I share your concerns about prematurely dropping support for these versions. Maintaining compatibility is important. I’ll ensure the updates remain compatible with Swift 5.7 and 5.8 for now.
2. Concurrency Checking: You’re right about leveraging full concurrency checking. This will help us identify potential issues early and maintain robustness in concurrent code. I’ll enable the `-Xfrontend -warn-concurrency` compiler flag and address any warnings that arise.
3. OpenID and RFC 8628 Scope: I understand your concerns regarding the scope of this PR, particularly the parts related to OpenID. My intention was to offer comprehensive enhancements, but I agree that keeping the scope focused is crucial. I’ll refactor the PR to remove the OpenID components, aligning it strictly with RFC 8628 requirements.

I’ll work on breaking down the PR into smaller, more manageable parts for a focused and effective review process.

**Comment by vamsii777** [comment]
@0xTim Gotcha, I'll break down the PR and YES would refactor so that you would be comfortable.

**Comment by vamsii777** [comment]
DO NOT MERGE. Moved to #28
